README.md
# aurelia-google-recaptcha

[![Npm Version](https://img.shields.io/npm/v/aurelia-google-recaptcha.svg)](https://www.npmjs.com/package/aurelia-google-recaptcha)
[![HitCount](http://hits.dwyl.io/dragoscirjan/aurelia-google-recaptcha.svg)](http://hits.dwyl.io/dragoscirjan/aurelia-google-recaptcha)
[![Contributions welcome](https://img.shields.io/badge/contributions-welcome-brightgreen.svg?style=flat)](https://github.com/dragoscirjan/aurelia-google-recaptcha/issues)

[![TravisCI](https://travis-ci.org/dragoscirjan/aurelia-google-recaptcha.svg?branch=master)](https://travis-ci.org/dragoscirjan/aurelia-google-recaptcha)
[![CircleCI](https://circleci.com/gh/dragoscirjan/aurelia-google-recaptcha.svg?style=shield)](https://circleci.com/gh/dragoscirjan/aurelia-google-recaptcha)

[![Donate to this project using Patreon](https://img.shields.io/badge/patreon-donate-yellow.svg)](https://patreon.com/dragoscirjan)
<!-- [![Donate to this project using Flattr](https://img.shields.io/badge/flattr-donate-yellow.svg)](https://flattr.com/profile/balupton)
[![Donate to this project using Liberapay](https://img.shields.io/badge/liberapay-donate-yellow.svg)](https://liberapay.com/dragoscirjan)
[![Donate to this project using Thanks App](https://img.shields.io/badge/thanksapp-donate-yellow.svg)](https://givethanks.app/donate/npm/badges)
[![Donate to this project using Boost Lab](https://img.shields.io/badge/boostlab-donate-yellow.svg)](https://boost-lab.app/dragoscirjan/badges)
[![Donate to this project using Buy Me A Coffee](https://img.shields.io/badge/buy%20me%20a%20coffee-donate-yellow.svg)](https://buymeacoffee.com/balupton)
[![Donate to this project using Open Collective](https://img.shields.io/badge/open%20collective-donate-yellow.svg)](https://opencollective.com/dragoscirjan)
[![Donate to this project using Cryptocurrency](https://img.shields.io/badge/crypto-donate-yellow.svg)](https://dragoscirjan.me/crypto)
[![Donate to this project using Paypal](https://img.shields.io/badge/paypal-donate-yellow.svg)](https://dragoscirjan.me/paypal)
[![Buy an item on our wishlist for us](https://img.shields.io/badge/wishlist-donate-yellow.svg)](https://dragoscirjan.me/wishlist) -->


An integration of [Google](https://google.com)'s [reCAPTCHA](https://developers.google.com/recaptcha/intro) api for [Aurelia](https://aurelia.io) Framework.

Plugin is inspired by [Jeremy Danyow](http://stackoverflow.com/users/725866/jeremy-danyow)'s [post](http://stackoverflow.com/questions/35441787/use-googles-recaptcha-in-an-aurelia-application), so please give a lot of credits to him as well, as we do.

## Index

<!-- TOC -->

- [aurelia-google-recaptcha](#aurelia-google-recaptcha)
    - [Index](#index)
    - [Audience](#audience)
    - [Getting Started](#getting-started)
    - [Usage](#usage)
        - [Generic Options](#generic-options)
    - [Overview](#overview)
    - [Development](#development)
    - [Issues](#issues)
    - [License](#license)

<!-- /TOC -->


## Audience

This documentation is designed for people familiar with [Aurelia](https://aurelia.io) Framework, HTML forms, server-side processing or mobile application development.

We hope you find this documentation easy to follow. For feedback and discussions please use the [issues](https://github.com/dragoscirjan/aurelia-google-recaptcha/issues) page of this project.

## Getting Started

```sh
npm i aurelia-google-recaptcha --save
```

or

```sh
yarn add aurelia-google-recaptcha
```

## Usage

In your aurelia `main.js` add:

```jsx
aurelia.use.plugin(PLATFORM.moduleName('aurelia-google-recaptcha'));
// or (if you want to configure the plugin from the start)
aurelia.use.plugin(PLATFORM.moduleName('aurelia-google-recaptcha'), config => {
    config.assign({
        siteKey: '', // use this as general sitekey if you use only one type of recaptcha
        siteKeys: {
          v2: '', // use these sitekeys to separte the values between all 3 types of recaptcha
          v2i: '',
          v3: ''
        },
        lang: 'en'  // for V2, we can also set the supported language
                    // see https://developers.google.com/recaptcha/docs/language
    });
});
```

In your HTML add one of the components as presented in the [Overview](#overview) section.

```html
<recaptcha-* ...></recaptcha-*>
```

### Generic Options

| Option | Type | Default | Description |
|---|---|---|---|
| `auto` | Boolean | `false` | _Optional._ Trigger an auto-validate loop interval, keeping the recaptcha validated. |
| `id` | String |   | _Optional._ _Two Way_ _bindable_ value, announcing recaptcha component ID. |
| `sitekey` | String  |   | _Optional._ Your can also mention the sitekey inline within code. This value will override the one from config. |
| `value` | String |   | _Optional._ _Two Way_ _bindable_ value, announcing recaptcha result outside of the component. |

## Overview

As [Google](https://google.com) will tell, you, to start using [reCAPTCHA](https://developers.google.com/recaptcha/intro), you need to sign up for an API key pair for your site. Please read the documentation on [reCAPTCHA](https://developers.google.com/recaptcha/intro) website to learn how to obtain the API keys.

From the versions presented by [reCAPTCHA](https://developers.google.com/recaptcha/intro), we chose to implement, within our [Aurelia](https://aurelia.io) plugin, the following:

* [reCAPTCHA v3](https://github.com/dragoscirjan/aurelia-google-recaptcha/blob/master/doc/recaptcha-v3.md)
* reCAPTCHA v2
  * [Checkbox](https://github.com/dragoscirjan/aurelia-google-recaptcha/blob/master/doc/recaptcha-v2.md)
  * [Invisible](https://github.com/dragoscirjan/aurelia-google-recaptcha/blob/master/doc/recaptcha-invisible.md)

For [verifying the user's response](https://developers.google.com/recaptcha/docs/verify) please read the reCAPTCHA documentation.

## Development

Fork the repo then clone it

`$ git clone git@github.com:dragoscirjan/aurelia-google-recaptcha-dev.git && cd aurelia-google-recaptcha`

`yarn global add gulp-cli` (or `$ npm install --global gulp-cli`): Install Gulp cli tool.

`$ yarn` (or `$ npm i`): Install the dependencies;

`$ yarn build` (or `$ npm run build`): Build to production;

`$ yarn dev`: Run the docz to see your changes;

> NOTE: Momentarely I do not use the testing environment. I'm using [this project](https://github.com/dragoscirjan/aurelia-plugins-dev) to develop. You're all invited to help with the unit tests, if you wish to.

`$ yarn test`: Run all tests: type checking and unit tests;

`$ yarn test:watch`: Watch unit tests;

`$ yarn tsc`: Typescript checking;

`$ yarn tsc:watch`: Typescript checking with watching;

## Issues

For issues, please reffer to this repository's [issue page](https://github.com/dragoscirjan/aurelia-google-recaptcha/issues).

## License

[MIT](https://github.com/dragoscirjan/aurelia-content-loader/blob/master/LICENSE)


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [0.0.3] - 2019-03-25
### Added
- Re-implemented [v2](https://developers.google.com/recaptcha/docs/display)
- Implemented [v2-invisible](https://developers.google.com/recaptcha/docs/invisible)
- Implemented [v3](https://developers.google.com/recaptcha/docs/v3)

## [0.0.1-2] - 2016-02-23
### Added
- 1st public release.

[Unreleased]: https://github.com/dragoscirjan/aurelia-content-loader/compare/v0.0.3...HEAD
[0.0.5]: https://github.com/dragoscirjan/aurelia-content-loader/compare/v0.0.1...v0.0.3
[0.0.4]: https://github.com/dragoscirjan/aurelia-content-loader/releases/tag/v0.0.1

test/unit/configure.spec.js
import {configure} from '../../src/index';

class ConfigStub {
  globalResources(...resources) {
    this.resources = resources;
  }
}

describe('the Aurelia configuration', () => {
  var mockedConfiguration;

  beforeEach(() => {
    mockedConfiguration = new ConfigStub();
    configure(mockedConfiguration);
  });

  it('should register a global resource', () => {
    expect(mockedConfiguration.resources).toContain('./recaptcha');
  });

});


src/config.js
/**
 * Aurelia Plugin :: Google Recaptcha (http://itmediaconnect.ro/aurelia/aurelia-google-recaptcha)
 *
 * @link      http://github.com/itmcdev/aurelia-google-re captcha for the canonical source repository
 * @link      https://github.com/ITMCdev/aurelia-google-recaptcha/issues for issues
 * @copyright Copyright (c) 2007-2016 IT Media Connect S.R.L. Romania (http://www.itmediaconnect.ro)
 * @license   https://github.com/ITMCdev/aurelia-google-recaptcha/blob/master/LICENSE MIT License
 */

import { extend } from 'amaranth-utils';
import { PLATFORM } from 'aurelia-pal';

export class Config {
  /** @var {} */
  values;

  /**
   * Constructor
   */
  constructor() {
    this.values = {
      siteKey: '',
      siteKeys: {
        v2: '',
        v2i: '',
        v3: ''
      },
      lang: 'en'
    };
  }

  /**
   * @param {FrameworkConfiguration} aurelia
   */
  configure(aurelia) {
    aurelia.globalResources(PLATFORM.moduleName('./recaptcha'));
    aurelia.globalResources(PLATFORM.moduleName('./recaptcha-v2'));
    aurelia.globalResources(PLATFORM.moduleName('./recaptcha-v2-invisible'));
  }

  /**
   * Get All
   * @returns {}
   */
  all() {
    return this.values;
  }

  /**
   * Assign
   * @param {{}} obj
   */
  assign(obj) {
    this.values = extend(this.values, obj);
  }

  /**
   * Getter
   * @param {String} key
   * @return {*}
   */
  get(key) {
    let result = extend({}, this.values);
    for (const chunk of key.split('.')) {
      result = result ? result[chunk] || undefined : undefined;
    }
    return result;
  }

  /**
   * Setter
   * @param {String} key
   * @param {*} value
   * @returns {*}
   */
  set(key, value) {
    chunks = key.split('.');
    if (chunks.length === 1) {
      this.values[key] = value;
      return this.values[key];
    }
    let part;
    while (chunks.length > 1) {
      part = part ? part[chunks.shift()] : this.values[chunks.shift()];
    }
    const lastKey = chunks.shift();
    part[lastKey] = value;
    return this.values[lastKey];
  }
}


src/recaptcha.js
/**
 * Aurelia Plugin :: Google Recaptcha (http://itmediaconnect.ro/aurelia/aurelia-google-recaptcha)
 *
 * @link      http://github.com/itmcdev/aurelia-google-re captcha for the canonical source repository
 * @link      https://github.com/ITMCdev/aurelia-google-recaptcha/issues for issues
 * @copyright Copyright (c) 2007-2016 IT Media Connect S.R.L. Romania (http://www.itmediaconnect.ro)
 * @license   https://github.com/ITMCdev/aurelia-google-recaptcha/blob/master/LICENSE MIT License
 */

import { bindable, customElement } from 'aurelia-framework';
import parseDuration from 'parse-duration';

import { GR, RecaptchaBase } from './recaptcha-base';

/**
 * Google Recaptcha plugin, originally developed by Jeremy Danyow (http://stackoverflow.com/users/725866/jeremy-danyow)
 * and extended by Dragos Cirjan (http://github.com/dragoscirjan)
 *
 * @link http://stackoverflow.com/questions/35441787/use-googles-recaptcha-in-an-aurelia-application
 * @link http://plnkr.co/edit/rXqNdh?p=info
 * @link https://developers.google.com/recaptcha/docs/v3
 *
 * @method noView
 * @method inject
 */
@customElement('recaptcha')
export class Recaptcha extends RecaptchaBase {
  /**
   * @var {String}
   * Values:
   * * homepage 	  See a cohesive view of your traffic on the admin console while filtering scrapers.
   * * login 	      With low scores, require 2-factor-authentication or email verification to prevent credential stuffing attacks.
   * * social 	    Limit unanswered friend requests from abusive users and send risky comments to moderation.
   * * e-commerce 	Put your real sales ahead of bots and identify risky transactions.
   * Note: actions may only contain alphanumeric characters and slashes, and must not be user-specific.
   */
  @bindable action = 'homepage';

  /** @var {Boolean} Optional. Trigger an auto-validate loop interval, keeping the recaptcha validated. */
  @bindable expires = '1m';

  /**
   * Component Bind Event
   */
  bind() {
    super.bind && super.bind();
    if (!this.sitekey) {
      this.sitekey = this.config.get('siteKeys.v3');
    }

    this.loadScript(this.getScriptId(), `https://www.google.com/recaptcha/api.js?render=${this.sitekey}`);

    this.registerExecuteEvent();
  }

  /**
   * Component Attached Event
   */
  attached() {
    super.attached && super.attached();

    this.registerExecuteWatcher();
    this.registerExpireWatcher();
  }

  /**
   * Create a watcher to invalidate recaptcha token based on `expires` time limit.
   */
  registerExpireWatcher() {
    const eventName = `grecaptcha:expire:${this.id}`;
    this.__watchers__[eventName] = setInterval(() => {
      if (this.lastCheck && new Date().getTime() - this.lastCheck.getTime() > this.parsedExpires) {
        this.value = null;
      }
    }, 1000);
  }

  /**
   * @returns {Number} Parsed duration in miliseconds.
   */
  get parsedExpires() {
    return parseDuration(this.expires);
  }

  /**
   * Call recaptcha.execute
   */
  recaptchaExecute() {
    if (typeof window[GR] === undefined || !grecaptcha.execute) {
      return;
    }
    grecaptcha.execute(this.sitekey, { action: this.action }).then(token => {
      this.value = token;
      this.lastCheck = new Date();
    });
  }
}


src/recaptcha-base.js
/** @global grecaptcha */
/**
 * Aurelia Plugin :: Google Recaptcha (http://itmediaconnect.ro/aurelia/aurelia-google-recaptcha)
 *
 * @link      http://github.com/itmcdev/aurelia-google-re captcha for the canonical source repository
 * @link      https://github.com/ITMCdev/aurelia-google-recaptcha/issues for issues
 * @copyright Copyright (c) 2007-2016 IT Media Connect S.R.L. Romania (http://www.itmediaconnect.ro)
 * @license   https://github.com/ITMCdev/aurelia-google-recaptcha/blob/master/LICENSE MIT License
 */

import { bindable, bindingMode, inject, noView } from 'aurelia-framework';
import { EventAggregator } from 'aurelia-event-aggregator';
import nanoid from 'nanoid';

import { Config } from './config';

/**
 * Hash method.
 * @returns {String}
 */
export const getHash = () => nanoid().replace(/-/gi, '');

/** @var {String} */
export const GR = 'grecaptcha';

@noView()
// @inject(Config, Element, EventAggregator)
@inject(Element, Config, EventAggregator)
export class RecaptchaBase {
  /** @var {Boolean} Optional. Trigger an auto-validate loop interval, keeping the recaptcha validated. */
  @bindable auto = false;

  /** @var {Boolean} Two way bindable value, announcing recaptcha component ID. */
  @bindable({ defaultBindingMode: bindingMode.twoWay }) id = null;

  /** @var {String} Your sitekey. */
  @bindable sitekey = '';

  /** @var {Boolean} Two way bindable value, announcing recaptcha result outside of the component. */
  @bindable({ defaultBindingMode: bindingMode.twoWay }) value = false;

  /** @var {String} */
  static HASH = getHash();

  /** @var {Config} */
  config;

  /** @var {String|null} The event name used to trigger the recaptcha validation. */
  eventName = null;

  /** @var {Date|null} The last time of grecaptcha.execute response.*/
  lastCheck = null;

  /** @var {{}} */
  __auevents__ = {};

  /** @var {{}} */
  __watchers__ = {};

  /**
   * Constructor
   * @param  {DocumentElement}    element Element to render recaptcha
   * @param  {EventAggregator}    events  Events Aggregator
   */
  constructor(element, config, events) {
    this.element = element;
    this.config = config;
    this.events = events;
  }

  /**
   * Component Bind Handler
   */
  bind() {
    if (!this.id) {
      this.id = getHash();
    }
    if (!this.sitekey) {
      this.sitekey = this.config.get('siteKey');
    }
  }

  /**
   * Component Detached Handler
   */
  detached() {
    // this.unloadScript(this.getScriptId());
    this.unregisterEvents();
    this.unregisterWatchers();
  }

  /**
   * Compose script id.
   * @returns {String}
   */
  getScriptId() {
    return `recaptcha_${RecaptchaBase.HASH}_${this.sitekey}`;
  }

  /**
   * Load a <script> by script id and src.
   * @param {String} id
   * @param {String} src
   */
  loadScript(id, src) {
    if (!document.querySelector(`#${id}`)) {
      let script = document.createElement('script');
      script.src = src;
      script.id = id;
      script.async = true;
      script.defer = true;
      document.head.appendChild(script);
    }
    this.element.setAttribute('data-script', this.getScriptId());
  }

  /**
   * Call recaptcha.execute
   */
  recaptchaExecute() {
    throw new Error('Method is to be implemented in each extended class.');
  }

  /**
   * Component Bind Execute Event
   */
  registerExecuteEvent() {
    const eventName = `grecaptcha:execute:${this.id}`;
    this.__auevents__[eventName] = this.events.subscribe(eventName, () => {
      this.recaptchaExecute();
    });
  }

  /**
   * Register an auto trigger for grecaptcha.execute using setInterval
   * @param {String} eventName
   */
  registerExecuteWatcher() {
    const eventName = `grecaptcha:execute:${this.id}`;
    this.__watchers__[eventName] = setInterval(() => {
      if (this.auto && !this.value) {
        this.events.publish(eventName);
      }
    }, 2000);
  }

  /**
   * Clear the interval for grecaptcha.execute
   */
  removeAutoTriggerExecuteEvent() {
    clearInterval(this.__grecaptcha_auto_exec_interval__);
  }

  /**
   * Unload a <script> by a given id
   * @param {String} id
   * @deprecated
   */
  unloadScript(id) {
    if (!document.querySelectorAll(`[data-script=${id}]`).length && document.querySelector(`#${id}`)) {
      document.querySelector(`#${id}`).remove();
    }
    if (!document.querySelectorAll('[data-script]').length && typeof window[GR] !== 'undefined') {
      delete window[GR];
    }
  }

  /**
   * Unregister an event.
   * @param {String} eventName
   */
  unregisterEvents() {
    for (const key of Object.keys(this.__auevents__)) {
      if (typeof this.__auevents__[key].dispose === 'function') {
        this.__auevents__[key].dispose();
        delete this.__auevents__[key];
      }
    }
  }

  /**
   * Unregister an event.
   * @param {String} eventName
   */
  unregisterWatchers() {
    for (const key of Object.keys(this.__watchers__)) {
      clearInterval(this.__watchers__[key]);
      delete this.__watchers__[key];
    }
  }
}


src/recaptcha-v2-invisible.js
/**
 * Aurelia Plugin :: Google Recaptcha (http://itmediaconnect.ro/aurelia/aurelia-google-recaptcha)
 *
 * @link      http://github.com/itmcdev/aurelia-google-re captcha for the canonical source repository
 * @link      https://github.com/ITMCdev/aurelia-google-recaptcha/issues for issues
 * @copyright Copyright (c) 2007-2016 IT Media Connect S.R.L. Romania (http://www.itmediaconnect.ro)
 * @license   https://github.com/ITMCdev/aurelia-google-recaptcha/blob/master/LICENSE MIT License
 */

import { bindable, customElement } from 'aurelia-framework';

import { RecaptchaV2 } from './recaptcha-v2';

/**
 * Google Recaptcha plugin, originally developed by Jeremy Danyow (http://stackoverflow.com/users/725866/jeremy-danyow)
 * and extended by Dragos Cirjan (http://github.com/dragoscirjan)
 *
 * @link https://developers.google.com/recaptcha/docs/invisible
 *
 * @method noView
 * @method inject
 */
@customElement('recaptcha-v2-invisible')
export class RecaptchaV2Invisible extends RecaptchaV2 {
  /** @var {String} Optional. Reposition the reCAPTCHA badge. 'inline' lets you position it with CSS. Values: bottomright, bottomleft, inline. Default: inline. */
  @bindable badge = 'inline';

  /** @var {Boolean} Optional. For plugin owners to not interfere with existing reCAPTCHA installations on a page. If true, this reCAPTCHA instance will be part of a separate ID space. */
  @bindable isolated = false;

  /**
   * Component Bind Event
   */
  bind() {
    super.bind && super.bind();
    if (!this.sitekey || this.sitekey === this.config.get('siteKeys.v2')) {
      this.sitekey = this.config.get('siteKeys.v2i');
    }

    this.registerExecuteEvent();
  }

  /**
   * Component Attached Event
   */
  attached() {
    super.attached && super.attached();

    this.registerExecuteWatcher();
  }

  /**
   * Call recaptcha.execute
   */
  recaptchaExecute() {
    grecaptcha.execute(this.widgetId);
  }

  renderOptions() {
    return {
      ...super.renderOptions(),
      badge: this.badge,
      isolated: this.isolated,
      size: 'invisible'
    };
  }
}


src/recaptcha-v2.js
/**
 * Aurelia Plugin :: Google Recaptcha (http://itmediaconnect.ro/aurelia/aurelia-google-recaptcha)
 *
 * @link      http://github.com/itmcdev/aurelia-google-re captcha for the canonical source repository
 * @link      https://github.com/ITMCdev/aurelia-google-recaptcha/issues for issues
 * @copyright Copyright (c) 2007-2016 IT Media Connect S.R.L. Romania (http://www.itmediaconnect.ro)
 * @license   https://github.com/ITMCdev/aurelia-google-recaptcha/blob/master/LICENSE MIT License
 */

import { bindable, customElement } from 'aurelia-framework';
import { getHash, RecaptchaBase } from './recaptcha-base';

let scriptReady = false;
const callbackName = `__recaptcha_callback_${getHash()}__`;
const callbackPromise = new Promise(
  resolve =>
    (window[callbackName] = () => {
      scriptReady = true;
      resolve();
    })
);

export class RecaptchaVerified /*extends Event*/ {
  constructor(component, token, widgetId) {
    // super('recaptcha-callback');

    this.component = component;
    this.grecaptcha = grecaptcha;
    this.token = token;
    this.widgetId = widgetId;
  }
}

export class RecaptchaError /*extends Event*/ {
  constructor(component, widgetId) {
    // super('recaptcha-callback-error');

    this.component = component;
    this.grecaptcha = grecaptcha;
    this.widgetId = widgetId;
  }
}

export class RecaptchaExpired /*extends Event*/ {
  constructor(component, widgetId) {
    // super('recaptcha-callback-expired');

    this.component = component;
    this.grecaptcha = grecaptcha;
    this.widgetId = widgetId;
  }
}

/**
 * Google Recaptcha plugin, originally developed by Jeremy Danyow (http://stackoverflow.com/users/725866/jeremy-danyow)
 * and extended by Dragos Cirjan (http://github.com/dragoscirjan)
 *
 * @link http://stackoverflow.com/questions/35441787/use-googles-recaptcha-in-an-aurelia-application
 * @link http://plnkr.co/edit/rXqNdh?p=info
 * @link https://developers.google.com/recaptcha/docs/display
 *
 * @method noView
 * @method inject
 */
@customElement('recaptcha-v2')
export class RecaptchaV2 extends RecaptchaBase {
  /** @var {Function|String} Optional. The name of your callback function, executed when the user submits a successful response. The g-recaptcha-response token is passed to your callback. */
  @bindable callback;

  /** @var {Function|String} Optional. The name of your callback function, executed when reCAPTCHA encounters an error (usually network connectivity) and cannot continue until connectivity is restored. If you specify a function here, you are responsible for informing the user that they should retry. */
  @bindable errorCallback;

  /** @var {Function|String} Optional. The name of your callback function, executed when the reCAPTCHA response expires and the user needs to re-verify. */
  @bindable expiredCallback;

  /** @var {String} Optional. The size of the widget. Values: compact, normal. Default: normal. */
  @bindable size = 'normal';

  /** @var {Number} Optional. The tabindex of the widget and challenge. If other elements in your page use tabindex, it should be set to make user navigation easier. */
  @bindable tabindex = 0;

  /** @var {String} Optional. The color theme of the widget. Values: dark, light. Default: light. */
  @bindable theme = 'light';

  /** @var {Number|null} ID oF widget, as per v2 of the recaptcha */
  widgetId = null;

  /**
   * Component Bind Handler
   */
  bind() {
    super.bind && super.bind();
    if (!this.sitekey) {
      this.sitekey = this.config.get('siteKeys.v2');
    }

    const lang = document.getElementsByTagName('html')[0].getAttribute('lang') || this.config.get('lang');
    const script = `https://www.google.com/recaptcha/api.js?onload=${callbackName}&render=explicit&hl=${lang}`;
    this.loadScript(this.getScriptId(), script);

    this.registerResetEvent();
  }

  /**
   * Component Attached Handler
   */
  attached() {
    super.attached && super.attached();

    this.render();
  }

  /**
   * Component Bind Reset Event
   */
  registerResetEvent() {
    const eventName = `grecaptcha:reset:${this.id}`;
    this.__auevents__[eventName] = this.events.subscribe(eventName, () => {
      grecaptcha && grecaptcha.reset && grecaptcha.reset(this.widgetId);
      this.value = null;
    });
  }

  /**
   * Render Recaptcha
   */
  render() {
    if (!scriptReady) {
      return callbackPromise.then(() => this.render());
    }
    this.widgetId = grecaptcha.render(this.element, this.renderOptions());
  }

  /**
   * Obtain Recaptcha Render Options
   * @returns {{}}
   */
  renderOptions() {
    const callback = () => {
      if (this.widgetId !== null) {
        this.value = grecaptcha.getResponse(this.widgetId);
      }
      if (this.callback) {
        if (typeof this.callback === 'function') {
          return this.callback(new RecaptchaVerified(this, this.value, this.widgetId));
        }
        if (window[this.callback]) {
          return window[this.callback].call(null, new RecaptchaVerified(this, this.value, this.widgetId));
        }
      }
    };
    const errorCallback = () => {
      this.value = undefined;
      if (this.errorCallback) {
        if (typeof this.errorCallback === 'function') {
          return this.callback(new RecaptchaError(this, this.widgetId));
        }
        if (window[this.errorCallback]) {
          return window[this.errorCallback].call(null, new RecaptchaError(this, this.widgetId));
        }
      }
    };
    const expiredCallback = () => {
      this.value = null;
      if (this.expiredCallback) {
        if (typeof this.expiredCallback === 'function') {
          return this.expiredCallback(new RecaptchaExpired(this, this.widgetId));
        }
        if (window[this.expiredCallback]) {
          return window[this.expiredCallback].call(null, new RecaptchaExpired(this, this.widgetId));
        }
      }
    };
    return {
      callback: callback,
      'error-callback': errorCallback,
      'expired-callback': expiredCallback,
      sitekey: this.sitekey,
      size: this.size,
      tabindex: this.tabindex,
      theme: this.theme,
      type: this.type
    };
  }
}
